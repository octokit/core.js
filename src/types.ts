import * as OctokitTypes from "@octokit/types";

import { Octokit } from ".";

export type RequestParameters = OctokitTypes.RequestParameters;

export type OctokitOptions = {
  // TODO: add types for authStrategy & auth options and octokit.auth() method,
  //       see https://tinyurl.com/typescript-auth-strategies
  authStrategy?: any;
  auth?: any;
  request?: OctokitTypes.RequestRequestOptions;
  timeZone?: string;
  [option: string]: any;
};

export type Constructor<T> = new (...args: any[]) => T;

export type SimplifyEmpty<T> = T extends void | undefined | null ? void : T

export type ReturnTypeOf<
  T extends AnyFunction | AnyFunction[]
> = T extends AnyFunction
  ? SimplifyEmpty<ReturnType<T>>
  : T extends AnyFunction[]
  ? UnionToIntersection<SimplifyEmpty<ReturnType<T[number]>>>
  : never;

/**
 * @author https://stackoverflow.com/users/2887218/jcalz
 * @see https://stackoverflow.com/a/50375286/10325032
 */
export type UnionToIntersection<Union> = (Union extends any
? (argument: Union) => void
: never) extends (argument: infer Intersection) => void // tslint:disable-line: no-unused
  ? Intersection
  : never;

type AnyFunction = (...args: any) => any;

export type OctokitPlugin = (
  octokit: Octokit,
  options: OctokitOptions
) => { [key: string]: any } | void | undefined;
