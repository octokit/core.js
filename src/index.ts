import { getUserAgent } from "universal-user-agent";
import { Collection, HookCollection } from "before-after-hook";
import { request } from "@octokit/request";
import { graphql, withCustomRequest } from "@octokit/graphql";
import { createTokenAuth } from "@octokit/auth-token";

import {
  Constructor,
  OctokitOptions,
  OctokitPlugin,
  RequestParameters,
  ReturnTypeOf,
  UnionToIntersection
} from "./types";
import { VERSION } from "./version";

export class Octokit {
  static VERSION = VERSION;
  static defaults<S extends Constructor<any>>(
    this: S,
    defaults: OctokitOptions
  ) {
    const OctokitWithDefaults = class extends this {
      constructor(...args: any[]) {
        const options = args[0] || {};

        super(
          Object.assign(
            {},
            defaults,
            options,
            options.userAgent && defaults.userAgent
              ? {
                  userAgent: `${options.userAgent} ${defaults.userAgent}`
                }
              : null
          )
        );
      }
    };

    return OctokitWithDefaults;
  }

  static plugins: OctokitPlugin[] = [];
  static plugin<
    S extends Constructor<any> & { plugins: any[] },
    T1 extends OctokitPlugin | OctokitPlugin[],
    T2 extends OctokitPlugin,
    T3 extends OctokitPlugin,
    T4 extends OctokitPlugin,
    T5 extends OctokitPlugin,
    T6 extends OctokitPlugin,
    T7 extends OctokitPlugin,
    T8 extends OctokitPlugin,
    T9 extends OctokitPlugin[]
  >(this: S, p1: T1, p2?: T2, p3?: T3, p4?: T4, p5?: T5, p6?: T6, p7?: T7, p8?: T8, ...p9: T9) {
    const currentPlugins = this.plugins;
      let newPlugins: (OctokitPlugin | undefined)[] = [
        ...(p1 instanceof Array ? p1 as OctokitPlugin[] : [p1 as OctokitPlugin]), 
        p2, p3, p4, p5, p6, p7, p8, ...p9
      ].map(v => {
        return typeof v === 'function' ? v : undefined
      })

    const NewOctokit = class extends this {
      static plugins = currentPlugins.concat(
        newPlugins.filter(plugin => !currentPlugins.includes(plugin))
      );
    };

    return NewOctokit as typeof NewOctokit
      & Constructor<UnionToIntersection<
          ReturnTypeOf<T1>
        & ReturnTypeOf<T2> 
        & ReturnTypeOf<T3> 
        & ReturnTypeOf<T4> 
        & ReturnTypeOf<T5> 
        & ReturnTypeOf<T6> 
        & ReturnTypeOf<T7> 
        & ReturnTypeOf<T8> 
        & ReturnTypeOf<T9>
      >>;
  }

  constructor(options: OctokitOptions = {}) {
    const hook = new Collection();
    const requestDefaults: Required<RequestParameters> = {
      baseUrl: request.endpoint.DEFAULTS.baseUrl,
      headers: {},
      request: Object.assign({}, options.request, {
        hook: hook.bind(null, "request")
      }),
      mediaType: {
        previews: [],
        format: ""
      }
    };

    // prepend default user agent with `options.userAgent` if set
    requestDefaults.headers["user-agent"] = [
      options.userAgent,
      `octokit-core.js/${VERSION} ${getUserAgent()}`
    ]
      .filter(Boolean)
      .join(" ");

    if (options.baseUrl) {
      requestDefaults.baseUrl = options.baseUrl;
    }

    if (options.previews) {
      requestDefaults.mediaType.previews = options.previews;
    }

    if (options.timeZone) {
      requestDefaults.headers["time-zone"] = options.timeZone;
    }

    this.request = request.defaults(requestDefaults);
    this.graphql = withCustomRequest(this.request).defaults(requestDefaults);
    this.log = Object.assign(
      {
        debug: () => {},
        info: () => {},
        warn: console.warn.bind(console),
        error: console.error.bind(console)
      },
      options.log
    );
    this.hook = hook;

    // (1) If neither `options.authStrategy` nor `options.auth` are set, the `octokit` instance
    //     is unauthenticated. The `this.auth()` method is a no-op and no request hook is registred.
    // (2) If only `options.auth` is set, use the default token authentication strategy.
    // (3) If `options.authStrategy` is set then use it and pass in `options.auth`. Always pass own request as many strategies accept a custom request instance.
    // TODO: type `options.auth` based on `options.authStrategy`.
    if (!options.authStrategy) {
      if (!options.auth) {
        // (1)
        this.auth = async () => ({
          type: "unauthenticated"
        });
      } else {
        // (2)
        const auth = createTokenAuth(options.auth as string);
        // @ts-ignore  ¯\_(ツ)_/¯
        hook.wrap("request", auth.hook);
        this.auth = auth;
      }
    } else {
      const auth = options.authStrategy(
        Object.assign(
          {
            request: this.request
          },
          options.auth
        )
      );
      // @ts-ignore  ¯\_(ツ)_/¯
      hook.wrap("request", auth.hook);
      this.auth = auth;
    }

    // apply plugins
    // https://stackoverflow.com/a/16345172
    const classConstructor = this.constructor as typeof Octokit;
    classConstructor.plugins.forEach(plugin => {
      Object.assign(this, plugin(this, options));
    });
  }

  // assigned during constructor
  request: typeof request;
  graphql: typeof graphql;
  log: {
    debug: (message: string, additionalInfo?: object) => any;
    info: (message: string, additionalInfo?: object) => any;
    warn: (message: string, additionalInfo?: object) => any;
    error: (message: string, additionalInfo?: object) => any;
    [key: string]: any;
  };
  hook: HookCollection;

  // TODO: type `octokit.auth` based on passed options.authStrategy
  auth: (...args: unknown[]) => Promise<unknown>;

  [key: string]: any;
}
