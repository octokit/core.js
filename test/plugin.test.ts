import { Octokit } from "../src";

const pluginFoo = () => {
  return { foo: "bar" };
};
const pluginBaz = () => {
  return { baz: "daz" };
};
const pluginQaz = () => {
  return { qaz: "naz" };
};

describe("Octokit.plugin()", () => {
  it("gets called in constructor", () => {
    const MyOctokit = Octokit.plugin(pluginFoo);
    const myClient = new MyOctokit();
    expect(myClient.foo).toEqual("bar");
  });

  it("supports multiple plugins", () => {
    const MyOctokit = Octokit.plugin(pluginFoo, pluginBaz, pluginQaz);
    const myClient = new MyOctokit();
    expect(myClient.foo).toEqual("bar");
    expect(myClient.baz).toEqual("daz");
    expect(myClient.qaz).toEqual("naz");
  });
  it("does not override plugins of original constructor", () => {
    const MyOctokit = Octokit.plugin(pluginFoo);
    const myClient = new MyOctokit();
    expect(myClient.foo).toEqual("bar");
    const octokit = new Octokit();
    expect(octokit).not.toHaveProperty("foo");
  });

  it("receives client options", () => {
    const MyOctokit = Octokit.plugin((octokit, options) => {
      expect(options).toStrictEqual({
        foo: "bar",
      });
    });
    new MyOctokit({ foo: "bar" });
  });

  it("does not load the same plugin more than once", () => {
    const myPlugin = (octokit: Octokit) => {
      if ("customKey" in octokit) {
        throw new Error("Boom!");
      }

      return {
        customKey: true,
      };
    };
    const MyOctokit = Octokit.plugin(myPlugin).plugin(myPlugin);
    expect(() => new MyOctokit()).not.toThrow();
  });

  it("supports chaining", () => {
    const MyOctokit = Octokit.plugin(pluginFoo)
      .plugin(pluginBaz)
      .plugin(pluginQaz);

    const myClient = new MyOctokit();
    expect(myClient.foo).toEqual("bar");
    expect(myClient.baz).toEqual("daz");
    expect(myClient.qaz).toEqual("naz");
  });
});
