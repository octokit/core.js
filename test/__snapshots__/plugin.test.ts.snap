// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Octokit.plugin() supports array of plugins and warns of deprecated usage 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Passing an array of plugins to Octokit.plugin() has been deprecated.
Instead of:
  Octokit.plugin([plugin1, plugin2, ...])
Use:
  Octokit.plugin(plugin1, plugin2, ...)",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
